doc/user-guide.md

```diff
@@ -432,54 +432,92 @@ deleted until you remove the finalizer (ie, after your cleanup logic has success
 The following is a snippet from the controller file under `pkg/controller/memcached/memcached_controller.go`
 
 ```Go
+
+const memcachedFinalizer = "finalizer.cache.example.com"
+
 func (r *ReconcileMemcached) Reconcile(request reconcile.Request) (reconcile.Result, error) {
- 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
- 	reqLogger.Info("Reconciling Memcached")
-
- 	// Fetch the Memcached instance
- 	memcached := &cachev1alpha1.Memcached{}
- 	err := r.client.Get(context.TODO(), request.NamespacedName, memcached)
- 	...
- 	// Check if the APP CR was marked to be deleted
- 	isMemcachedMarkedToBeDeleted := memcached.GetDeletionTimestamp() != nil
- 	if isMemcachedMarkedToBeDeleted {
- 		// TODO(user): Add the cleanup steps that the operator needs to do before the CR can be deleted
- 		// Update finalizer to allow delete CR
- 		memcached.SetFinalizers(nil)
-
- 		// Update CR
- 		err := r.client.Update(context.TODO(), memcached)
- 		if err != nil {
- 			return reconcile.Result{}, err
- 		}
- 		return reconcile.Result{}, nil
- 	}
-
- 	// Add finalizer for this CR
- 	if err := r.addFinalizer(reqLogger, instance); err != nil {
- 		return reconcile.Result{}, err
- 	}
- 	...
-
- 	return reconcile.Result{}, nil
- }
-
-//addFinalizer will add this attribute to the Memcached CR
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Memcached")
+
+	// Fetch the Memcached instance
+	memcached := &cachev1alpha1.Memcached{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, memcached)
+
+	...
+
+	// Check if the Memcached instance is marked to be deleted
+	isMemcachedMarkedToBeDeleted := memcached.GetDeletionTimestamp() != nil
```

Comment:
Is there a potential race condition here? For example, `deletionTimestamp < time.Now()` so the object has already been deleted and `r.finalizeMemcached()` will fail on each reconciliation. Is there a better way to check if `memcached` was deleted, or is this an accepted practice?

Reply:
I am confused on what would be the race condition. 

This is how the ansible operator handles resources that have finalizers on them, can you explain a little more how `< time.Now()` would impact this logic?

Reply:
This is also how the helm operator works. I think the key concept is that if there's a deletion timestamp, it means there has been a request to delete the object, but the object has not yet been deleted because it still has finalizers.

If the object has actually been deleted, the original `r.client.Get` will fail with a not found error. And the operator developer should likely just return without error from the `Reconcile` function. Would it be more clear if I added that extra error check to the example?

Reply:
@joelanford I wasn't aware that the presence of a deletion timestamp implies that info, thanks! Yes, an extra check with a comment would be clearer.